### PR TITLE
fix: add validation for duplicate translations

### PR DESCRIPTION
**Pull request by vietnguyen**
https://github.com/dhis2/dhis2-core/pull/9459
Issue:
- User can post duplicate  translation records with same `property` and `locale` but different value.

Fix:
- Add new class`TranslationsCheck`. It will be used by `api/metadata` and also `api/{object-type}/translations`

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseIdentifiableObject.java [hunk 1]
```diff
@@ -802,4 +781,24 @@ public String toString()
...
+     * Get Translation value from {@code Set<Translation>} by given locale and
+     * translationKey
+     *
+     * @return Translation value if exists, otherwise return null.
+     */
+    private String getTranslationValue( String locale, String translationKey )
+    {
+        for ( Translation translation : translations )
+        {
+            if ( locale.equals( translation.getLocale() ) && translationKey.equals( translation.getProperty() ) &&
```
Comment (jbee) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/BaseIdentifiableObject.java:795]: I always wonder if this should not take care of the locale fallback, for example if you ask for `en_GB` but we only have `en` then looking up `en_GB` should end up using `en` when there is no `en_GB`.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/feedback/ObjectReport.java [hunk 2]
```diff
@@ -226,7 +227,8 @@ public boolean hasErrorReport( Predicate<ErrorReport> test )
     @Override
     public void forEachErrorReport( Consumer<ErrorReport> reportConsumer )
     {
-        errorReportsByCode.values().forEach( reports -> reports.forEach( reportConsumer ) );
+        errorReportsByCode.values().stream().flatMap( errors -> errors.stream() )
+            .collect( Collectors.toList() ).forEach( errorReport -> reportConsumer.accept( errorReport ) );
```
Comment (jbee) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/feedback/ObjectReport.java:231]: `.collect( Collectors.toList() ).forEach(` appears unnecessarily creating a intermediate collection just to enter stream processing again.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/validation/TranslationsCheck.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+import org.hisp.dhis.feedback.ErrorCode;
+import org.hisp.dhis.feedback.ErrorReport;
+import org.hisp.dhis.feedback.ObjectReport;
+import org.hisp.dhis.importexport.ImportStrategy;
+import org.hisp.dhis.schema.Schema;
+import org.hisp.dhis.translation.Translation;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author viet@dhis2.org
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/validation/TranslationsCheck.java:48]: Please state the purpose of the check in the javadoc - this is often quite hard to reverse engineer after some time :D

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/validation/TranslationsCheck.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+                    .setErrorKlass( klass ) );
+            }
+
+            if ( translation.getValue() == null )
+            {
+                objectReport.addErrorReport(
+                    new ErrorReport( Translation.class, ErrorCode.E4000, "value" ).setErrorKlass( klass ) );
+            }
+
+            addReports.accept( objectReport );
```
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/validation/TranslationsCheck.java:128]: Do we always want to add the report even if there is no error?

Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/validation/TranslationsCheck.java]: no, i added an empty check.

**Review by jbee — APPROVED** [review]

**Pull request merged**